### PR TITLE
Fix `tt.dot` codegen when input operands are integers

### PR DESCRIPTION
**Pull request by etiotto**
Fix codegen for `tt.dot` so that:
  - operands that do not have the same type as the result type are promoted correctly
  - FMA instruction is generated only when the operands have floating point type
  - LLVM mul + add instructions are generated when the operands have integer type

Fixes issue #191.

### CODE REVIEW

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMPass.cpp [hunk 1]
```diff
@@ -919,10 +920,29 @@ struct ConvertTritonGPUToLLVM
...
-    return builder.create<triton::FpToFpOp>(loc, tensorPromotedType, operand);
+    Type elemType = tensorPromotedType.getElementType();
+    return llvm::TypeSwitch<Type, Value>(elemType)
+        .Case<FloatType>([&](auto) {
+          assert(isa<FloatType>(elemType) && "Expecting a floating point type");
+          return builder.create<triton::FpToFpOp>(loc, tensorPromotedType,
+                                                  operand);
+        })
+        .Case<IntegerType>([&](auto) {
+          assert(isa<IntegerType>(elemType) && "Expecting an integer type");
```
Comment (whitneywhtsang) [inline lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMPass.cpp:934]:
```suggestion
```

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMPass.cpp [hunk 2]
```diff
@@ -919,10 +920,29 @@ struct ConvertTritonGPUToLLVM
...
                               Type promotedType) {
-    Type tensorPromotedType =
+    auto tensorPromotedType =
         operand.getType().cast<RankedTensorType>().cloneWith(std::nullopt,
                                                              promotedType);
-    return builder.create<triton::FpToFpOp>(loc, tensorPromotedType, operand);
+    Type elemType = tensorPromotedType.getElementType();
+    return llvm::TypeSwitch<Type, Value>(elemType)
+        .Case<FloatType>([&](auto) {
+          assert(isa<FloatType>(elemType) && "Expecting a floating point type");
```
Comment (whitneywhtsang) [inline lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMPass.cpp:929]:
```suggestion
```

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMPass.cpp [hunk 3]
```diff
@@ -919,10 +920,29 @@ struct ConvertTritonGPUToLLVM
 
   static Value promoteOperand(OpBuilder &builder, Location loc, Value operand,
                               Type promotedType) {
-    Type tensorPromotedType =
+    auto tensorPromotedType =
```
Comment (whitneywhtsang) [inline lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMPass.cpp:923]:
```suggestion
    Type tensorPromotedType =
```

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMPass.cpp (hunk 1 above)
Comment (etiotto) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMPass.cpp:934]: done

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMPass.cpp (hunk 2 above)
Comment (etiotto) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMPass.cpp:929]: done

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMPass.cpp (hunk 3 above)
Comment (etiotto) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMPass.cpp:923]: Using `Type` would require a cast to `RankedTensorType` to be able to call `getElementType()` on the next line. Leaving as `auto`.

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**